Update TS/JS toggle button design

aomarks opened the pull request:
Trying to make it look more interactive. Open to suggestions/tweaks.

### New design
<img src="https://user-images.githubusercontent.com/48894/132930316-a61f7e33-ea9b-4895-af1c-391a30af08f1.png" width="500px">

### New design hover
<img src="https://user-images.githubusercontent.com/48894/132930327-3e2642cc-2855-406b-bba6-504edcadbc2c.png" width="500px">

### Old design

<img src="https://user-images.githubusercontent.com/48894/132930409-ca22da53-a7ae-467d-ba44-64cc1d9ae765.png" width="500px">

AndrewJakubowicz reviewed (commented):
Haven't gone through code yet. Just some quick initial feedback.
 - It looks amazing! Tried with keyboard controls and it works well.
 - Tested on Windows: Firefox & Chrome

Chrome suggests a slightly larger contrast between the white text and background:
![image](https://user-images.githubusercontent.com/15080861/132932112-3baf12a1-f1a9-4939-8c4b-6786459f5988.png)

AndrewJakubowicz commented on packages/lit-dev-content/src/components/litdev-code-language-switch.ts:
```diff
@@ -25,45 +24,70 @@ export class LitDevCodeLanguageSwitch extends LitElement {
+    }
+
+    [position='js'] > #toggle {
+      left: 0;
```
elegant!

AndrewJakubowicz commented on packages/lit-dev-content/src/components/litdev-code-language-switch.ts:
```diff
@@ -25,45 +24,70 @@ export class LitDevCodeLanguageSwitch extends LitElement {
+      width: 50%;
+      height: 100%;
+      top: 0;
+      transition: left 100ms;
```
Not required but maybe could add [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)? You can sometimes get many moving toggles on the same view.

aomarks replied:
Cool, TIL. Done.

AndrewJakubowicz commented on packages/lit-dev-content/src/components/litdev-code-language-switch.ts:
```diff
@@ -93,33 +117,20 @@ export class LitDevCodeLanguageSwitch extends LitElement {
+        role="switch"
+        aria-checked=${mode == 'ts' ? 'true' : 'false'}
+        aria-title="Toggle TypeScript"
+        title=${mode === 'ts' ? 'Disable TypeScript' : 'Enable TypeScript'}
```
No action required nit: Should the titles be more along the lines of `Toggle between TypeScript and JavaScript`, rather than enabling/disabling TypeScript?

aomarks replied:
> Chrome suggests a slightly larger contrast between the white text and background:

Fixed!

aomarks replied:
> No action required nit: Should the titles be more along the lines of `Toggle between TypeScript and JavaScript`, rather than enabling/disabling TypeScript?

Good question, not 100% sure about this but ... the reason I used "Toggle TypeScript" for the `aria-label` is that it seemed like `role=switch` was the best role to use, which means that the state of the button is expressed in the `aria-checked` attribute. It seemed like "Toggle between TypeScript and JavaScript" would be confusing in that case, because it's not clear whether TypeScript or JavaScript maps to true vs false. When using voiceover, this reads as "Toggle TypeScript" and then "on"/"off" as you switch between them. 

And for `title`, it also seemed good to make it the same boolean choice for visual users, too.

AndrewJakubowicz reviewed (approved):
This looks fantastic! LGTM

justinfagnani commented:
This looks great!

A couple of design notes:

- The vertical alignment needs a small adjustment to make it visually centered I think. I added a 1px bottom padding to the <button> which helped.
- The horizontal centering of the toggle could maybe use an adjustment too. I see there's a .25em left padding, which could be bumped up slightly (1px, but in em I suppose) on the JS label, but reduced slightly on the TS label.
- The toggle drop shadow makes it look a little blurry to me, especially in the gray / non-hover state. Maybe drop the opacity to 20%? Another option is to only show the drop shadows on hover.
- Maybe the button could have a transparent background, then white on hover?

aomarks commented:
> * The vertical alignment needs a small adjustment to make it visually centered I think. I added a 1px bottom padding to the  which helped.
> * The horizontal centering of the toggle could maybe use an adjustment too. I see there's a .25em left padding, which could be bumped up slightly (1px, but in em I suppose) on the JS label, but reduced slightly on the TS label.

I made some changes to help with centering, including some new padding. PTAL.

Also, I'm trying a technique with units that just occurred to me. Pick some default `font-size`, and then use `calc(<px>em / <default-font-size>)` everywhere else, where `<px>` is the pixel measurement you want at the default font size. That way you can easily write pixel-perfect units for the default size, but still allow arbitrary scaling.

> * The toggle drop shadow makes it look a little blurry to me, especially in the gray / non-hover state. Maybe drop the opacity to 20%? Another option is to only show the drop shadows on hover.

I removed all the drop shadows (also the inset one). Definitely crisper, I think I prefer this look. WDYT?

> * Maybe the button could have a transparent background, then white on hover?

I think anything we do with opacity would drop the contrast so that it's no longer accessible. The white-on-gray is already the minimum contrast that passes Chrome's a11y check.

justinfagnani commented:
> I think anything we do with opacity would drop the contrast so that it's no longer accessible. The white-on-gray is already the minimum contrast that passes Chrome's a11y check.

Are there any a11y guidelines for hover-over effects like this? Like, does the non-hovered state have to have the same contrast as the hovered state? Certainly if we had decide to only show the switch at all on hover, the contrast of the non-hovered state wouldn't be an issue.

aomarks commented:
> Are there any a11y guidelines for hover-over effects like this? Like, does the non-hovered state have to have the same contrast as the hovered state? Certainly if we had decide to only show the switch at all on hover, the contrast of the non-hovered state wouldn't be an issue.

Good question. I would think it would be important for the non-hover state to meet contrast guidelines -- otherwise how would somebody who couldn't perceive it clearly even know to hover over it? And lighthouse would definitely dock our score regardless.
